File: COMP1020/InClassCode/Week3Code/file_reading_demo.py
"""
FileReadingDemo.py

This module demonstrates different ways to read files in Python,
mirroring the Java examples for educational comparison.

Differences between Java and Python are noted in comments.
"""
import sys

def main(args):
    filename = "file.txt"

    print("--- Method 1: open() read(1) (Char by Char) ---")
    read_with_file_reader(filename)
    print("\n")

    print("--- Method 2: open() readline() (Line by Line) ---")
    read_with_buffered_reader(filename)
    print("\n")

    print("--- Method 3: open() iterator (Token/Line) ---")
    read_with_scanner(filename)
    print("\n")

def read_with_file_reader(filename):
    """
    Method 1: "FileReader" equivalent
    READS: One character at a time.

    DIFFERENCES:
    - Java: Use FileReader class.
    - Python: Use built-in open() function.
    - Java: read() returns an int (-1 at EOF).
    - Python: read(1) returns a string of length 1 (empty string '' at EOF).
    - Error Handling: Java uses Checked Exceptions. Python uses Unchecked Exceptions.
    """
    # Python uses try-except blocks, not try-catch
    try:
        # open() is the built-in function to access files.
        # 'r' mode means read-only (default).
        fr = open(filename, 'r')

        # Read one character (passing 1 to read())
        char = fr.read(1)

        # Check for empty string which indicates EOF in Python
        while char != '':
            # print() adds a newline by default; use end='' to suppress it
            # to match the behavior of Java's System.out.print
            print(char, end='')
            char = fr.read(1)

        # Always close the stream!
        fr.close()

    except IOError as e:
        # Python catches IOError (or OSError) for file related issues
        print(f"An error occurred: {e}")

def read_with_buffered_reader(filename):
    """
    Method 2: "BufferedReader" equivalent
    READS: Whole lines at a time.

    DIFFERENCES:
    - Java: Must wrap FileReader in BufferedReader to get readLine().
    - Python: The file object returned by open() already has buffering
      and a readline() method.
    """
    try:
        # No wrapper needed in Python; open() returns a file object
        br = open(filename, 'r')

        line = br.readline()

        # readline() returns empty string '' at end of file.
        # Unlike Java's readLine() which returns null.
        while line != '':
            # Java's readLine() strips the newline character, so println adds one.
            # Python's readline() keeps the newline character.
            # To match Java's output, we allow proper spacing or control it with end=''.
            print(line, end='')

            line = br.readline()

        br.close()

    except IOError as e:
        print(f"An error occurred: {e}")

def read_with_scanner(filename):
    """
    Method 3: "Scanner" equivalent
    READS: Lines (or tokens) by iterator.

    DIFFERENCES:
    - Java: Scanner is a distinct class for parsing tokens/lines.
    - Python: Iterating directly over the file object is the standard
      "line-by-line" reading method, similar to Scanner's hasNextLine loop.
    """
    try:
        file_obj = open(filename, 'r')

        # Standard for-loop pattern for files in Python.
        # This is equivalent to while(sc.hasNextLine())
        for line in file_obj:
            print(line, end='')

        file_obj.close()

    except FileNotFoundError as e:
        # Scanner specifically throws FileNotFoundException (subclass of IOException)
        # Python raises FileNotFoundError
        print(f"File not found: {filename}")

# Standard boilerplate to run the main method
if __name__ == "__main__":
    main(sys.argv)

File: COMP1020/InClassCode/Week3Code/FileReadingDemo.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

public class FileReadingDemo {

    public static void main(String[] args) {
        String filename = "file.txt";

        System.out.println("--- Method 1: FileReader (Char by Char) ---");
        readWithFileReader(filename);
        System.out.println("\n");

        System.out.println("--- Method 2: BufferedReader (Line by Line) ---");
        readWithBufferedReader(filename);
        System.out.println("\n");

        System.out.println("--- Method 3: Scanner (Token/Line) ---");
        readWithScanner(filename);
        System.out.println("\n");
    }

    /**
     * Method 1: FileReader
     * READS: One character at a time.
     * PROS: Simple for small data or character-level processing.
     * CONS: Very inefficient (many disk accesses) and tedious to reconstruct strings.
     */
    public static void readWithFileReader(String filename) {
        // FileReader uses "Checked Exceptions", so we must try-catch IOException
        try {
            FileReader fr = new FileReader(filename);
            int i;
            // .read() returns -1 when end of file is reached
            i = fr.read();
            while (i != -1) {
                // Cast the integer value to a char to print it
                System.out.print((char) i);
                i = fr.read();
            }
            // Always close the stream!
            fr.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Method 2: BufferedReader
     * READS: Whole lines at a time.
     * PROS: Efficient (buffers input), easy to read text line-by-line.
     * CONS: Slightly more code to set up (wrapping the FileReader).
     */
    public static void readWithBufferedReader(String filename) {
        try {
            // BufferedReader wraps around FileReader to provide buffering
            BufferedReader br = new BufferedReader(new FileReader(filename));

            String line = br.readLine();
            // .readLine() returns null when end of file is reached
            while (line != null) {
                System.out.println(line);

                line = br.readLine();
            }

            br.close(); // Closing the wrapper also closes the inner FileReader
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Method 3: Scanner
     * READS: Tokens (words, numbers) or Lines.
     * PROS: easiest to parse data (nextInt, nextDouble), very high-level.
     * CONS: Slower than BufferedReader for massive files. Throws unchecked exceptions during parsing.
     */
    public static void readWithScanner(String filename) {
        try {
            File file = new File(filename);
            Scanner sc = new Scanner(file);

            // standard while-loop pattern for Scanner
            while (sc.hasNextLine()) {
                String line = sc.nextLine();
                System.out.println(line);
            }

            sc.close();
        } catch (FileNotFoundException e) {
            // Scanner specifically throws FileNotFoundException (subclass of IOException)
            // It suppresses general IOExceptions during reading and returns null/false instead.
            System.out.println("File not found: " + filename);
        }
    }
}
